### PR TITLE
impulse cost on CoM

### PR DESCRIPTION
**Pull request by proyan**
Implements the cost for impulse effect on com.
This PR still lacks the unittests. What is the current standard for unittesting?
Solves #285

### CODE REVIEW

File: bindings/python/crocoddyl/multibody/impulse-costs/com-position.hpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+//
+// Copyright (C) 2018-2019, LAAS-CNRS
+// Copyright note valid unless otherwise stated in individual files.
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef BINDINGS_PYTHON_CROCODDYL_MULTIBODY_IMPULSE_COSTS_COM_POSITION_HPP_
+#define BINDINGS_PYTHON_CROCODDYL_MULTIBODY_IMPULSE_COSTS_COM_POSITION_HPP_
+
+#include "crocoddyl/multibody/impulse-costs/com-position.hpp"
```
Comment (cmastalli) [inline bindings/python/crocoddyl/multibody/impulse-costs/com-position.hpp]: Move this file to "crocoddyl/multibody/impulse-com-position.hpp"

File: bindings/python/crocoddyl/multibody/impulse-costs/com-position.hpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+          bp::args("self", "state"),
+          "Initialize the CoM position cost model.\n\n"
+          "For this case the default activation model is quadratic, i.e.\n"
+          "crocoddyl.ActivationModelQuad(3), and nu is equals to model.nv.\n"
+          ":param state: state of the multibody system"))
+      .def("calc", &ImpulseCostModelCoM::calc_wrap, bp::args("self", "data", "x"),
+           "Compute the CoM position cost.\n\n"
+           ":param data: cost data\n"
+           ":param x: time-discrete state vector")
+      .def<void (ImpulseCostModelCoM::*)(const boost::shared_ptr<ImpulseCostDataAbstract>&, const Eigen::VectorXd&,
```
Comment (cmastalli) [inline bindings/python/crocoddyl/multibody/impulse-costs/com-position.hpp]: You will need to adapt the bindings. For instance, the code will depend on CostDataAbstract

File: bindings/python/crocoddyl/multibody/impulse-costs/com-position.hpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+namespace crocoddyl {
+namespace python {
+
+namespace bp = boost::python;
+
+void exposeImpulseCostCoM() {
+  bp::class_<ImpulseCostModelCoM, bp::bases<ImpulseCostModelAbstract> >(
+      "ImpulseCostModelCoM", bp::init<boost::shared_ptr<StateMultibody>, boost::shared_ptr<ActivationModelAbstract> >(
+                                 bp::args("self", "state", "activation"),
+                                 "Initialize the CoM position cost model.\n\n"
```
Comment (cmastalli) [inline bindings/python/crocoddyl/multibody/impulse-costs/com-position.hpp]: Improve documentation, e.g. "Initialize the CoM position cost model under impulse dynamics.\n\n

File: include/crocoddyl/multibody/impulse-cost-base.hpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
+///////////////////////////////////////////////////////////////////////////////
```
Comment (cmastalli) [inline include/crocoddyl/multibody/impulse-cost-base.hpp]: Again, you need to remove this file

File: include/crocoddyl/multibody/impulse-costs/com-position.hpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (C) 2018-2020, LAAS-CNRS, University of Edinburgh
+// Copyright note valid unless otherwise stated in individual files.
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef CROCODDYL_MULTIBODY_IMPULSE_COSTS_COM_POSITION_HPP_
```
Comment (cmastalli) [inline include/crocoddyl/multibody/impulse-costs/com-position.hpp]: Adapt the guard according the new path of the file

File: include/crocoddyl/multibody/impulse-costs/com-position.hpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+#define CROCODDYL_MULTIBODY_IMPULSE_COSTS_COM_POSITION_HPP_
+
+#include "crocoddyl/multibody/impulse-cost-base.hpp"
+#include "crocoddyl/multibody/data/impulses.hpp"
+#include "crocoddyl/core/utils/exception.hpp"
+
+namespace crocoddyl {
+
+class ImpulseCostModelCoM : public ImpulseCostModelAbstract {
```
Comment (cmastalli) [inline include/crocoddyl/multibody/impulse-costs/com-position.hpp]: Inherent from CostModelAbstract

File: include/crocoddyl/multibody/impulse-costs/com-position.hpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+ public:
+  ImpulseCostModelCoM(boost::shared_ptr<StateMultibody> state, boost::shared_ptr<ActivationModelAbstract> activation);
+  ImpulseCostModelCoM(boost::shared_ptr<StateMultibody> state);
+  ~ImpulseCostModelCoM();
+
+  void calc(const boost::shared_ptr<ImpulseCostDataAbstract>& data, const Eigen::Ref<const Eigen::VectorXd>& x);
```
Comment (cmastalli) [inline include/crocoddyl/multibody/impulse-costs/com-position.hpp]: And this depends on CostDataAbstract

File: include/crocoddyl/multibody/impulse-costs/com-position.hpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+                const bool& recalc = true);
+  boost::shared_ptr<ImpulseCostDataAbstract> createData(DataCollectorAbstract* const data);
+};
+
+struct ImpulseCostDataCoM : public ImpulseCostDataAbstract {
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  template <typename Model>
+  ImpulseCostDataCoM(Model* const model, DataCollectorAbstract* const data)
+      : ImpulseCostDataAbstract(model, data),
```
Comment (cmastalli) [inline include/crocoddyl/multibody/impulse-costs/com-position.hpp]: CostDataAbstract

File: include/crocoddyl/multibody/impulse-costs/com-position.hpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        Arr_Rx(3, model->get_state()->get_nv()),
+        dvc_dq(3, model->get_state()->get_nv()),
+        ddv_dv(model->get_state()->get_nv(), model->get_state()->get_nv()),
+        pinocchio_dv(pinocchio::Data(model->get_state()->get_pinocchio())) {
+    Arr_Rx.fill(0);
+    dvc_dq.fill(0);
+    ddv_dv.fill(0);
+    // Check that proper shared data has been passed
+    DataCollectorMultibody* d = dynamic_cast<DataCollectorMultibody*>(shared);
```
Comment (cmastalli) [inline include/crocoddyl/multibody/impulse-costs/com-position.hpp]: Use DataCollectorImpulse

File: src/CMakeLists.txt [hunk 10]
```diff
@@ -36,6 +36,8 @@ SET(${PROJECT_NAME}_SOURCES
   multibody/states/multibody.cpp
   multibody/actuations/floating-base.cpp
   multibody/actuations/full.cpp
+  multibody/impulse-cost-base.cpp
```
Comment (cmastalli) [inline src/CMakeLists.txt]: Remove this file

File: src/CMakeLists.txt [hunk 11]
```diff
@@ -36,6 +36,8 @@ SET(${PROJECT_NAME}_SOURCES
...
+  multibody/impulse-costs/com-position.cpp
```
Comment (cmastalli) [inline src/CMakeLists.txt]: Move this file to "multibody/impulse-com-position.hpp"

File: src/multibody/impulse-cost-base.cpp [hunk 12]
```diff
@@ -0,0 +1,35 @@
+///////////////////////////////////////////////////////////////////////////////
```
Comment (cmastalli) [inline src/multibody/impulse-cost-base.cpp]: Not needed this file

File: src/multibody/impulse-costs/com-position.cpp [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+///////////////////////////////////////////////////////////////////////////////
+
+#include "crocoddyl/core/utils/exception.hpp"
+#include "crocoddyl/multibody/impulse-costs/com-position.hpp"
+#include <pinocchio/algorithm/center-of-mass.hpp>
+#include <pinocchio/algorithm/center-of-mass-derivatives.hpp>
+
+namespace crocoddyl {
+
+ImpulseCostModelCoM::ImpulseCostModelCoM(boost::shared_ptr<StateMultibody> state,
```
Comment (cmastalli) [inline src/multibody/impulse-costs/com-position.cpp]: Rename the class to "CostModelImpulseCoM"

File: src/multibody/impulse-costs/com-position.cpp [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+                 << "nr is equals to 3");
+  }
+}
+
+ImpulseCostModelCoM::ImpulseCostModelCoM(boost::shared_ptr<StateMultibody> state)
+    : ImpulseCostModelAbstract(state, 3) {}
+
+ImpulseCostModelCoM::~ImpulseCostModelCoM() {}
+
+void ImpulseCostModelCoM::calc(const boost::shared_ptr<ImpulseCostDataAbstract>& data,
```
Comment (cmastalli) [inline src/multibody/impulse-costs/com-position.cpp]: Adapt it, now it depends on CostDataAbstract

File: src/multibody/impulse-costs/com-position.cpp [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+  if (recalc) {
+    calc(data, x);
+  }
+
+  ImpulseCostDataCoM* d = static_cast<ImpulseCostDataCoM*>(data.get());
+
+  // Compute the derivatives of the frame placement
+  const std::size_t& nv = state_->get_nv();
+  const std::size_t& ndx = state_->get_ndx();
+  activation_->calcDiff(data->activation, data->r, recalc);
```
Comment (cmastalli) [inline src/multibody/impulse-costs/com-position.cpp]: activation_->calcDiff(data->activation, data->r, false);

File: src/multibody/impulse-costs/com-position.cpp [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+  pinocchio::getCenterOfMassVelocityDerivatives(state_->get_pinocchio(), d->pinocchio_dv, d->dvc_dq);
+  pinocchio::jacobianCenterOfMass(state_->get_pinocchio(), d->pinocchio_dv, false);
+  data->Rx.leftCols(nv) = d->dvc_dq + d->pinocchio_dv.Jcom * data->impulses->dvnext_dx.leftCols(nv);
```
Comment (cmastalli) [inline src/multibody/impulse-costs/com-position.cpp]: Split this Eigen expression as:
```cpp
data->Rx.leftCols(nv) = d->dvc_dq;
data->Rx.leftCols(nv).noalias() += d->pinocchio_dv.Jcom * data->impulses->dvnext_dx.leftCols(nv);
```

**Comment by proyan** [comment]
> Impulse-based cost functions needs to be inheritance by cost-base class (i.e. CostModelAbstract) as other cost functions.
I agree that we need to inherit from the CostModelAbstract. However, we also need an ImpulseModelAbstract. Normally the inheritance should be CostModelAbstract->CostModelImpulseAbstract->ImpulseModelCoM. The reason is to collect the binding to ImpulseData and ContactData, that we need to do in the case of Impulse and contact cost classes, into a single class.

Unfortunately, I couldn't inherit from the CostModelAbstract class, since the API of calc and calcdiff in impulse-based classes is `calc(x)`, and not `calc(x,u)`. In the CostModelAbstract class, calc(x) is not a virtual function, so I cannot simply avoid using calc(x,u) as I would have normally done if I  inherit from CostModelAbstract. If you are okay with making calc(x) as virtual, and moving the implemetation to the derived cost functions, this would be easier.

>  The trick is to parse the data smartly as done here: https://github.com/loco-3d/crocoddyl/blob/devel/include/crocoddyl/multibody/costs/contact-force.hpp#L43.

Currently you are re-doing the same implementation for every cost function that uses a contact-force. We need to have a CostModelContactForceAbstract class.

**Comment by cmastalli** [comment]
> Unfortunately, I couldn't inherit from the CostModelAbstract class, since the API of calc and calcdiff in impulse-based classes is `calc(x)`, and not `calc(x,u)`. In the CostModelAbstract class, calc(x) is not a virtual function, so I cannot simply avoid using calc(x,u) as I would have normally done if I inherit from CostModelAbstract. If you are okay with making calc(x) as virtual, and moving the implemetation to the derived cost functions, this would be easier.

calc(x) is not virtual because it calls calc(x,u) with u=np.zero. Therefore, you need to implement the calc/calcDiff functions on the impulse using this interface.

It wouldn't work this code, we need to use cost-base class.


> Currently you are re-doing the same implementation of the binding with ContactData for every costmodel class that uses a contact-force. We need to have a CostModelContactForceAbstract class, where we can do this implementation once.

No, we just need to parse the code as done already.
Please follow the contact-force implementation.

**Comment by cmastalli** [comment]
I will handle the changes, please do not perform any action

**Comment by cmastalli** [comment]
@proyan 

Thank you for starting this.
Now the code is ready to be merged. I did the following modifications:
 1. The impulse-com cost inheritance the cost-base class (necessary step for integration to the action models)
 2. Removed the impulse-com-base classes, not needed.
 3. Renamed the files to impulse-com
 4. Fixed a bug in the dimension of `ddv_dv`
 5. Parsed the impulse data from action "impulse" model (as done for contact costs)
 6. Created Python bindings of the impulse-com data.
 7. Added Python unit-test against action numdiff.

As soon it passed the CI, I will integrate it.
I will do the impulse friction cone soon.

Thanks again!

**Pull request merged**